javascriptlib/src/mill/javascriptlib/PublishModule.scala

```diff
@@ -0,0 +1,539 @@
+      .replaceFirst(project, "typescript")) ++ pubModDepsGenSources().map(pr =>
+      "typescript/generatedSources/" + pr.path.last
+    )
+
+  }
+
+  override def generatedSourcesPathsBuilder: T[Seq[(String, String)]] = Task {
+    Seq("@generated/*" -> "typescript/generatedSources")
+  }
+
+  override def upstreamPathsBuilder: Task[Seq[(String, String)]] = Task.Anon {
+
+    val upstreams = (for {
+      (res, mod) <- Task.traverse(moduleDeps)(_.resources)().zip(moduleDeps)
+    } yield {
+      Seq((
+        mod.millSourcePath.subRelativeTo(Task.workspace).toString + "/*",
+        s"typescript/${mod.millSourcePath.subRelativeTo(Task.workspace).toString}/src" + ":" + s"${declarationOut()}"
+      )) ++
+        res.map { rp =>
+          val resourceRoot = rp.path.last
+          val modName = mod.millSourcePath.subRelativeTo(Task.workspace).toString
+          // nb: resources are be moved in bundled stage
+          (
+            "@" + modName + s"/$resourceRoot/*",
+            resourceRoot match {
+              case s if s.contains(".dest") =>
+                rp.path.toString
+              case _ => s"typescript/$modName/$resourceRoot"
+            }
+          )
+        }
+
+    }).flatten
+
+    upstreams
+  }
+
+  override def typeRoots: T[ujson.Value] = Task.Anon {
+    ujson.Arr(
+      "node_modules/@types",
+      "declarations"
+    )
+  }
+
+  override def declarationDir: T[ujson.Value] = Task.Anon {
+    ujson.Str("declarations")
+  }
+
+  override def compilerOptionsPaths: Task[Map[String, String]] =
+    Task.Anon { Map.empty[String, String] }
+
+  override def compilerOptions: T[Map[String, ujson.Value]] = Task {
+    Map(
+      "declarationMap" -> ujson.Bool(true),
+      "esModuleInterop" -> ujson.Bool(true),
+      "baseUrl" -> ujson.Str("."),
+      "rootDir" -> ujson.Str("typescript"),
+      "declaration" -> ujson.Bool(true),
+      "outDir" -> ujson.Str(bundledOut()),
+      "plugins" -> ujson.Arr(
+        ujson.Obj("transform" -> "typescript-transform-paths"),
+        ujson.Obj(
+          "transform" -> "typescript-transform-paths",
+          "afterDeclarations" -> true
+        )
+      ),
+      "moduleResolution" -> ujson.Str("node"),
+      "module" -> ujson.Str("CommonJS"),
+      "target" -> ujson.Str("ES2020")
+    )
+  }
+
+  // patch typescript
+  private def tsPatchInstall: T[Unit] = Task {
+    os.call(
+      ("node", npmInstall().path / "node_modules/ts-patch/bin/ts-patch", "install"),
+      cwd = npmInstall().path
+    )
+    ()
+  }
+
+  override def symlinkNodeModules: Task[Unit] = Task {
+    tsPatchInstall() // patch typescript compiler => use custom transformers
+    os.call(
+      ("ln", "-s", npmInstall().path.toString + "/node_modules/", "node_modules"),
+      cwd = publishDir().path
+    )
+    if (os.exists(npmInstall().path / ".npmrc")) os.call(
+      ("ln", "-s", npmInstall().path.toString + "/.npmrc", ".npmrc"),
+      cwd = publishDir().path
```

Comment:
These should be using `os.symlink` rather than `os.call`

Reply:
couldn't get it to work, an error about symlink already existing kept popping up.
I was unable to resolve it, defaulted back to `os.call`.

Reply:
Let's figure this out and either use `os.symlink` or file a minimized ticket as to why it's not working. `os.*` APIs are there for us to use, and we should use them where possible over subprocesses